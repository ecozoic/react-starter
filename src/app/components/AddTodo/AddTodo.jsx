import React from 'react';
import PropTypes from 'prop-types';
import { Button, Input, Icon } from 'semantic-ui-react';

class AddTodo extends React.Component {
  static propTypes = {
    onInit: PropTypes.func.isRequired,
    onSubmit: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);

    this.state = {
      todo: '',
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    // TODO: only fetch if not initialized
    this.props.onInit();
  }

  handleSubmit(event) {
    const { onSubmit } = this.props;
    const { todo } = this.state;

    event.preventDefault();

    if (!todo || !todo.trim()) {
      return;
    }

    onSubmit(todo);

    this.setState({ todo: '' });
  }

  handleChange(event) {
    this.setState({ todo: event.target.value });
  }

  render() {
    const { todo } = this.state;

    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <Input
            name="todo"
            type="text"
            value={todo}
            onChange={this.handleChange}
            action={
              <Button
                type="submit"
                primary
                animated
                disabled={!todo}
              >
                <Button.Content visible>
                  Add Todo
                </Button.Content>
                <Button.Content hidden>
                  <Icon name="plus" />
                </Button.Content>
              </Button>
            }
          />
        </form>
      </div>
    );
  }
}

export default AddTodo;
