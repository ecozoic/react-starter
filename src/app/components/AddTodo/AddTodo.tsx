import React from 'react';
import PropTypes from 'prop-types';
import { Button, Input, Icon } from 'semantic-ui-react';

export interface AddTodoProps {
  onInit: () => void;
  onSubmit: (todo: string) => void;
}

export interface AddTodoState {
  todo: string;
}

class AddTodo extends React.Component<AddTodoProps, AddTodoState> {
  static propTypes = {
    onInit: PropTypes.func.isRequired,
    onSubmit: PropTypes.func.isRequired,
  };

  constructor(props: AddTodoProps) {
    super(props);

    this.state = {
      todo: '',
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    // TODO: only fetch if not initialized
    this.props.onInit();
  }

  handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    const { onSubmit } = this.props;
    const { todo } = this.state;

    event.preventDefault();

    if (!todo || !todo.trim()) {
      return;
    }

    onSubmit(todo);

    this.setState({ todo: '' });
  }

  handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ todo: event.target.value });
  }

  render() {
    const { todo } = this.state;

    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <Input
            name="todo"
            type="text"
            value={todo}
            onChange={this.handleChange}
            action={
              <Button
                type="submit"
                primary
                animated
                disabled={!todo}
              >
                <Button.Content visible>
                  Add Todo
                </Button.Content>
                <Button.Content hidden>
                  <Icon name="plus" />
                </Button.Content>
              </Button>
            }
          />
        </form>
      </div>
    );
  }
}

export default AddTodo;
