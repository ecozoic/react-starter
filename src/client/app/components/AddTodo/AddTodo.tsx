import * as React from 'react';
import { Component, FormEvent, ChangeEvent } from 'react';
import * as PropTypes from 'prop-types';
import RaisedButton from 'material-ui/RaisedButton';
import TextField from 'material-ui/TextField';

interface AddTodoProps {
  onInit: () => void;
  onSubmit: (todo: string) => void;
}

interface AddTodoState {
  todo: string;
}

class AddTodo extends Component<AddTodoProps, AddTodoState> {
  static propTypes = {
    onInit: PropTypes.func.isRequired,
    onSubmit: PropTypes.func.isRequired,
  };

  constructor(props: AddTodoProps) {
    super(props);

    this.state = {
      todo: '',
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    // TODO: only fetch if not initialized
    this.props.onInit();
  }

  handleSubmit(event: FormEvent<HTMLFormElement>) {
    const { onSubmit } = this.props;
    const { todo } = this.state;

    event.preventDefault();

    if (!todo || !todo.trim()) {
      return;
    }

    onSubmit(todo);

    this.setState({ todo: '' });
  }

  handleChange(event: ChangeEvent<HTMLInputElement>) {
    this.setState({ todo: event.target.value });
  }

  render() {
    const { todo } = this.state;

    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <TextField
            name="todo"
            type="text"
            value={todo}
            onChange={this.handleChange}
          />
          <RaisedButton
            label="Add Todo"
            primary
            type="submit"
          />
        </form>
      </div>
    );
  }
}

export default AddTodo;
