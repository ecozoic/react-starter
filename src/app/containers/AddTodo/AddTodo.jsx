import { connect } from 'react-redux';

import { addTodo, fetchTodos } from '../../actions/index';
import AddTodo from '../../components/AddTodo/index';

const mapDispatchToProps = dispatch => ({
  onSubmit: (todo) => {
    dispatch(addTodo(todo));
  },
  onInit: () => {
    dispatch(fetchTodos());
  },
});

const AddTodoContainer = connect(
  undefined,
  mapDispatchToProps,
)(AddTodo);

export default AddTodoContainer;
