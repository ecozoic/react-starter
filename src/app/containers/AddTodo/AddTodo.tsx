import { Dispatch } from 'redux';
import { connect } from 'react-redux';

import { State } from '../../reducers';
import { addTodo, fetchTodos } from '../../actions';
import { getPending, getError } from '../../selectors';
import AddTodo from '../../components/AddTodo';

const mapStateToProps = (state: State) => ({
  disabled: getPending(state) || !!getError(state),
});

const mapDispatchToProps = (dispatch: Dispatch<State>) => ({
  onSubmit: (todo: string) => {
    dispatch(addTodo(todo));
  },
  onInit: () => {
    dispatch(fetchTodos());
  },
});

const AddTodoContainer = connect<any, any, any>(
  mapStateToProps,
  mapDispatchToProps,
)(AddTodo);

export default AddTodoContainer;
