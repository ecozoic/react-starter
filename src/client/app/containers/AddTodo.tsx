import { Dispatch } from 'redux';
import { connect } from 'react-redux';

import { State } from '../reducers';
import { addTodo, fetchTodos } from '../actions';
import AddTodo from '../components/AddTodo';

const mapDispatchToProps = (dispatch: Dispatch<State>) => ({
  onSubmit: (todo: string) => {
    dispatch(addTodo(todo));
  },
  onInit: () => {
    dispatch(fetchTodos());
  },
});

const AddTodoContainer = connect(
  undefined,
  mapDispatchToProps,
)(AddTodo);

export default AddTodoContainer;
