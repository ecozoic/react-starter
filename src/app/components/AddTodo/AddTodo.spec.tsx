import React from 'react';
import { mount, shallow } from 'enzyme';

import { noop } from '../../utils';

import AddTodo from './';

describe('<AddTodo />', () => {
  it('renders without crashing', () => {
    const addTodo = shallow(
      <AddTodo
        onInit={noop}
        onSubmit={noop}
      />,
    );

    expect(addTodo).toMatchSnapshot();
  });

  describe('handleSubmit()', () => {
    it('prevents default behavior on form submission', () => {
      const preventDefault = jest.fn();

      const event: any = {
        preventDefault,
      };

      const addTodo = shallow(
        <AddTodo
          onInit={noop}
          onSubmit={noop}
        />,
      );

      const form = addTodo.find('form');
      form.simulate('submit', event);

      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it('noops for empty and whitespace strings', () => {
      const onSubmit = jest.fn();

      const addTodo = shallow(
        <AddTodo
          onInit={noop}
          onSubmit={onSubmit}
        />,
      );

      addTodo.setState({ todo: '' });

      const instance = addTodo.instance() as AddTodo;
      const event: any = { preventDefault: noop };

      instance.handleSubmit(event);

      expect(onSubmit).not.toHaveBeenCalled();

      addTodo.setState({ todo: '    ' });

      instance.handleSubmit(event);

      expect(onSubmit).not.toHaveBeenCalled();
    });

    it('calls onSubmit for valid todos', () => {
      const onSubmit = jest.fn();
      const todo = 'todo';

      const addTodo = shallow(
        <AddTodo
          onInit={noop}
          onSubmit={onSubmit}
        />,
      );

      addTodo.setState({ todo });

      const instance = addTodo.instance() as AddTodo;
      const event: any = { preventDefault: noop };

      instance.handleSubmit(event);

      expect(onSubmit).toHaveBeenCalledTimes(1);
      expect(onSubmit).toHaveBeenCalledWith(todo);
    });

    it('resets state after submission', () => {
      const addTodo = shallow(
        <AddTodo
          onInit={noop}
          onSubmit={noop}
        />,
      );

      addTodo.setState({ todo: 'foo' });

      const instance = addTodo.instance() as AddTodo;
      const event: any = { preventDefault: noop };

      instance.handleSubmit(event);

      expect(addTodo.state('todo')).toEqual('');
    });
  });

  describe('handleChange()', () => {
    it('sets state', () => {
      const addTodo = shallow(
        <AddTodo
          onInit={noop}
          onSubmit={noop}
        />,
      );

      const instance = addTodo.instance() as AddTodo;
      const todo = 'todo';
      const event: any = { target: { value: todo } };

      instance.handleChange(event);

      expect(addTodo.state('todo')).toEqual(todo);
    });
  });

  describe('componentDidMount', () => {
    it('calls onInit after mounting', () => {
      const onInit = jest.fn();

      mount(
        <AddTodo
          onInit={onInit}
          onSubmit={noop}
        />,
      );

      expect(onInit).toHaveBeenCalledTimes(1);
    });
  });
});
