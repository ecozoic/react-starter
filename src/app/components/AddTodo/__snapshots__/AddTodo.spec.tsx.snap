// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddTodo /> renders without crashing 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <Input
      name="todo"
      onChange={[Function]}
      type="text"
      value=""
    />
    <Button
      as="button"
      primary={true}
      type="submit"
    >
      Add Todo
    </Button>
  </form>
</div>
`;
